### PR TITLE
[SelectionDAG] Salvage debug info for non-constant ADDs

### PR DESCRIPTION
**Pull request by dstenb**
Teach SelectionDAG::salvageDebugInfo() to salvage debug information for
ADD nodes where the RHS is non-constant.
    
Co-authored-by: Mikael Holmen <mikael.holmen@ericsson.com>

- [DebugInfo] Precommit testcase for pointer addition with unknown offset
- [SelectionDAG] Salvage debug info for non-constant ADDs

### CODE REVIEW

**Comment by dstenb** [comment]
Fix clang-format issue.

File: llvm/test/DebugInfo/Sparc/pointer-add-unknown-offset-debug-info.mir [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  ; CHECK: bb.0.entry:
+  ; CHECK-NEXT:   liveins: $i0, $i1
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:i64regs = COPY $i1
+  ; CHECK-NEXT:   [[COPY1:%[0-9]+]]:i64regs = COPY $i0
+  ; CHECK-NEXT:   [[SRAri:%[0-9]+]]:i64regs = SRAri [[COPY]], 0
+  ; CHECK-NEXT:   [[SLLXri:%[0-9]+]]:i64regs = SLLXri killed [[SRAri]], 2
+  ; CHECK-NEXT:   DBG_VALUE_LIST !13, !DIExpression(DW_OP_LLVM_arg, 0, DW_OP_LLVM_arg, 1, DW_OP_plus, DW_OP_stack_value), [[COPY1]], [[SLLXri]], debug-location !16
+  ; CHECK-NEXT:   DBG_VALUE_LIST !14, !DIExpression(DW_OP_LLVM_arg, 0, DW_OP_LLVM_arg, 1, DW_OP_plus, DW_OP_plus_uconst, 3, DW_OP_stack_value), [[COPY1]], [[SLLXri]], debug-location !16
+  ; CHECK-NEXT:   DBG_VALUE_LIST !15, !DIExpression(DW_OP_LLVM_arg, 0, DW_OP_LLVM_arg, 2, DW_OP_plus, DW_OP_LLVM_arg, 1, DW_OP_LLVM_arg, 3, DW_OP_plus, DW_OP_plus, DW_OP_stack_value), [[COPY1]], [[COPY1]], [[SLLXri]], [[SLLXri]], debug-location !16
```
Comment (adrian-prantl) [inline llvm/test/DebugInfo/Sparc/pointer-add-unknown-offset-debug-info.mir:57]: Is this testing the non-variadic case, too?

**Review by adrian-prantl — APPROVED** [review]

File: llvm/test/DebugInfo/Sparc/pointer-add-unknown-offset-debug-info.mir (hunk 1 above)
Comment (dstenb) [reply llvm/test/DebugInfo/Sparc/pointer-add-unknown-offset-debug-info.mir:57]: Yes, the debug values for "p" and "q" are still non-variadic when entering `SelectionDAG::salvageDebugInfo()`:

```
 DbgVal(Order=3)(SDNODE=t9:0):"p"
 DbgVal(Order=3)(SDNODE=t9:0):"q"!DIExpression(DW_OP_plus_uconst, 3, DW_OP_stack_value)
```

**Comment by dstenb** [comment]
Reverted in 6b25890130955ee6ac5127c43c835e3ef6c6850b.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -10791,27 +10795,46 @@ void SelectionDAG::salvageDebugInfo(SDNode &N) {
...
+            NewLocOps.push_back(RHS);
+            DIExpr = DIExpression::appendOpsToArg(TmpDIExpr, ExprOps, i, true);
+          }
           Changed = true;
         }
         (void)Changed;
         assert(Changed && "Salvage target doesn't use N");
 
+        bool IsVariadic =
+            DV->isVariadic() || OrigLocOpsSize != NewLocOps.size();
```
Comment (KanRobert) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:10832]: We changed the value of `IsVariadic` here and `getDbgValueList`(line 10835) would call constructor of `SDDbgValue`
https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/SelectionDAG/SDNodeDbgValue.h#L166

it seems the assertion `assert(!(IsVariadic && IsIndirect));` in the constructor may fail.  In fact, I did see such an error in a downstream project.

Comment (KanRobert) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:10832]: https://github.com/llvm/llvm-project/issues/72630

**Pull request merged**